db/column_family.cc

```diff
@@ -670,7 +668,11 @@ bool ColumnFamilyData::UnrefAndTryDelete() {
     return true;
   }
 
-  if (old_refs == 2 && super_version_ != nullptr) {
+  // If called under SuperVersion::Cleanup, we should not re-enter Cleanup on
+  // the same SuperVersion. (But while installing a new SuperVersion, this
+  // cfd could be referenced only by two SuperVersions.)
+  if (old_refs == 2 && super_version_ != nullptr &&
+      super_version_ != sv_under_cleanup) {
```

Comment:
If `sv_under_cleanup` is only used by `SuperVersion::Cleanup()`, should this logic also move to there? I feel having a SuperVersion a parameter for `UnrefAndTryDelete` is a little confusing.

Reply:
When sv_under_cleanup is non-null, we might exclude this code. We might need the code below from any caller. If you have a specific proposal to simplify, let me know. It is not simple as it is here.

Reply:
I think my suggestion is to change `SuperVersion::Cleanup()`, like:
```
  if (cfd->GetSuperVersion() == this) {
    if (cfd->Unref()) {
      delete cfd;
    }
  } else {
    cfd->UnrefAndTryDelete();
  }
```

Reply:
We would either need to keep the Unref() function around, which I fear invites further bugs, or make SuperVersion a friend of ColumnFamilyData. Making a `friend` suggests we would be violating the natural encapsulation for ColumnFamilyData.